design/nfs/nfs-provisioner-controlled-by-operator.md

```diff
@@ -0,0 +1,82 @@
+---
+title: nfs-provisioner-controlled-by-operator
+target-version: release-X.X
+---
+
+# NFS Provisioner Controlled by Operator
+
+## Summary
+
+Rook NFS Provisioner is a built in dynamic provisioner for Rook NFS. The functionality works fine but has a bug which the provisioner uses the same directory between the provisioned PV.
```

Comment:
I think i'm missing something fundamental for this issue, and the mapping between NFSServer, shares/exports, and PVs.  That mapping (it doesn't have to be visual, it could just be textual) could perhaps make this issue and design more clear.

If a user creates a PVC to get a PV that is backed by a NFS export (share)...why is it a concern that another PVC/PV pointing to that same export is backed by the same directory in the provisioner?  NFS shares are read write many, so wouldn't you want them to all be sharing the same underlying data?

The mapping between all these concepts could help clear this up, or please add more background/info so this issue gets more clear about what the problem is and how this design fixes it. Thanks!

Reply:
> If a user creates a PVC to get a PV that is backed by a NFS export (share)...why is it a concern that another PVC/PV pointing to that same export is backed by the same directory in the provisioner? NFS shares are read write many, so wouldn't you want them to all be sharing the same underlying data?

I think every PV (that is backed by NFS export and provisioned by our provisioner) should have its own data. The problem is when the user creates PV (through rook nfs dynamic provisioner) called PV-A and PV-B. Both of them use the same directory in the `nfs-server export`. Both PV-A and PV-B will have the same data. I mean if the user wants to share data, the user can just use that PVC instead of creating another PVC/PV.

In another case, we will add a deletion feature in our provisioner (which currently not implemented yet). The user creates two PV (from NFS export) called PV-A and PV-B. If both PV-A and PV-B use the same directory for their data and the user deletes the PV-A then what will happen with PV-B? Also about `disk-quota` how provisioner can provide specific storage capacity for each PV while its in one directory?

Reply:
I see, I was thinking of each user requested PV to map to the root of the share/export that is backing it.  But we're suggesting in this design that instead, each user requested PV should be an independent sub-directory within the share/export.  They are both backed by the same share/export, but they are independent sub-directories that do not overlap.  

Is this hierarchy correct?

* underlying PV (could be Ceph, HostPath, etc.)
  * NFS share/export
    * PV-A provisioned for user 1
    * PV-B provisioned for user 2

In that hierarchy, PV-A and PV-B are backed by the same NFS share/export, but they are independent of each other and do not have any overlap or shared data. Is that correct? 

Please @ mention me when you're updated this design doc. A visual to make these mappings/hierarchy clear would probably be very helpful and effective.

Reply:
> In that hierarchy, PV-A and PV-B are backed by the same NFS share/export, but they are independent of each other and do not have any overlap or shared data. Is that correct?

Yup, that's correct.